tools/transform.cpp

```diff
@@ -379,6 +383,46 @@ static void calculateAndInitConstants(Transform &t) {
       return false;
     return true;
   };
+  // Returns access size. 0 if no access, -1 if unknown
+  const uint64_t UNKNOWN = -1, NO_ACCESS = 0;
+  auto get_access_size = [&](const Instr &inst) -> uint64_t {
+    if (dynamic_cast<const Alloc *>(&inst) ||
+        dynamic_cast<const Malloc *>(&inst))
+      // They are uninitialized (no actual bytes written), so assume that they
+      // are not accessed.
+      // Calloc is different: it contains memset to zero
+      return NO_ACCESS;
+
+    Type *value_ty = nullptr;
+    unsigned align = 0;
+    if (auto st = dynamic_cast<const Store *>(&inst)) {
+      value_ty = &st->getValue()->getType();
+      align = st->getAlign();
+    } else if (auto ld = dynamic_cast<const Load *>(&inst)) {
+      value_ty = &ld->getType();
+      align = ld->getAlign();
+    }
+
+    if (value_ty) {
+      if (auto ity = dynamic_cast<const IntType *>(value_ty))
+        return gcd(align, util::divide_up(ity->bits(), 8));
+      else if (auto fty = dynamic_cast<const FloatType *>(value_ty))
+        return gcd(align, fty->bits() / 8);
+      else if (dynamic_cast<const PtrType *>(value_ty))
+        return gcd(align, bits_size_t / 8);
+      // Aggregate types should consider padding
+      return UNKNOWN;
+    }
+
+    if (dynamic_cast<const BinOp *>(&inst) ||
```

Comment:
I think it would be clearer if all the no access instructions would be moved up.